<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/google/meridian/blob/main/demo/Meridian_Data_Guide-Meridian_Data_Basics.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/google/meridian/blob/main/demo/Meridian_Data_Guide-Meridian_Data_Basics.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
</table>

# **Meridian Data Guide: Meridian Data Basics**


<b>Table of Contents</b>
<ul>
    <li><a href="#introduction">Introduction: Meridian Data Basics</a></li>
    <li><a href="#key-input">Meridian's Key Data Input: Strategic & Privacy-Safe 🔐</a></li>
    <li><a href="#philosophy">Core Philosophy: Geo-Level Modeling & Data Volume</a>
        <ul>
            <li><a href="#national-structure">National-Level Data Structure</a></li>
            <li><a href="#geo-structure">Geo-Level Data Structure</a></li>
        </ul>
    </li>
    <li><a href="#schema">The Meridian Input Schema</a>
        <ul>
            <li><a href="#key-variables">A Closer Look at Key Input Variables</a>
                <ul>
                    <li><a href="#media-metrics">Media Metrics</a></li>
                    <li><a href="#control-variables">Control Variables</a></li>
                    <li><a href="#outcome-metrics">Outcome Metrics</a></li>
                    <li><a href="#geo-variables">Geo-Specific Variable</a></li>
                </ul>
            </li>
        </ul>
    </li>
    <li><a href="#samples">Sample Meridian Input Files</a></li>
</ul>

<a id="introduction"></a>
## **Introduction: Meridian Data Basics**

[Google Meridian](https://developers.google.com/meridian) is a powerful open-source [Marketing Mix Model (MMM)](https://en.wikipedia.org/wiki/Marketing_mix_modeling) framework from Google that helps you measure and optimize the impact of your marketing channels.

This guide covers the basic concepts of Meridian's data requirements, as the accuracy of any MMM depends entirely on the quality and structure of its input data. We will explore the ideal data structure, the different types of variables the model can handle, and what a finished, model-ready dataset looks like.

**Target Audience:** Marketers, data analysts, and business stakeholders who are new to preparing data for sophisticated marketing mix models, but is also geared towards technical data scientists looking for concrete examples.

**A Note on the Code 🐍**

This guide uses Python, but you do not need to be an expert programmer or statistician. The code is just a tool for the heavy lifting. All code snippets will be simplified and commented on.

Everything is designed to run in Google Colab, which works directly in your web browser - there’s no installation or complex setup required. You can focus on understanding your data & answering key marketing questions.

In [ ]:
# Import Python libraries required to get this notebook to run
import pandas as pd


<a id="key-input"></a>
## **Meridian's Key Data Input: Strategic & Privacy-Safe 🔐**

Meridian is designed to use **tabular panel data**. The structure is not unlike what you would see in a **typical Excel spreadsheet**, just loaded into a [Pandas DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) so Meridian can run advanced statistical models on it. This approach is fundamental to how the model works and is a major benefit for modern marketers.

* **Tabular:** The data is organized in a table with rows and columns, just like a spreadsheet.
* **Panel:** You consistently observe the same set of groups (e.g., the 50 US states) over many time periods (e.g., 104 weeks).

This approach is a deliberate choice because it is both strategically powerful and inherently respectful of user privacy.


### **Why It's Strategic**

Modeling on aggregated panel data allows you to focus on the big picture—the true goal of Marketing Mix Modeling. Instead of getting lost in the noise of individual user journeys, you can answer critical, high-level business questions. By observing how your business outcomes (like total conversions) change across different geographical regions with varying levels of media spend each week, the model can isolate the **true incremental impact** of your marketing. This approach is durable and doesn't rely on tracking individual users via cookies, making it a stable, future-proof measurement solution.

### **Why It's Privacy-Safe by Design**

This is a key benefit: tabular panel data contains **no Personally Identifiable Information (PII)**.

Because all the input data is **aggregated** to a high level (e.g., total weekly spend and conversions for the entire state of California), any link to individual users is removed. The model only sees that "California had 1,500 conversions and $80,000 in spend," not that a specific person saw an ad and then made a purchase. This focus on group-level activity ensures that your marketing measurement is effective without compromising user privacy.

### **What does it look like?**

Here is a small sample what it may look like. As you can see, no sensitive PII is involved in the data set:


In [ ]:
# Stores the web address (URL) of the raw CSV data file in a variable named 'url'.
url = "https://raw.githubusercontent.com/google/meridian/main/meridian/data/simulated_data/csv/geo_media_rf.csv"

# Uses the pandas library (pd) to read the data from the URL and load it into a DataFrame named 'df'.
df = pd.read_csv(url)

# Displays the first 5 rows of the DataFrame ('df.head(5)') to give a quick preview of the data structure and its contents.
display(df.head(5))

geo        time  Channel0_impression  Channel1_impression  \
0  Geo0  2021-01-25            1392518.0               3733.0   
1  Geo0  2021-02-01             937228.0             722210.0   
2  Geo0  2021-02-08            1286569.0             329778.0   
3  Geo0  2021-02-15            1149907.0             529628.0   
4  Geo0  2021-02-22            1077028.0            1057061.0   

   Channel2_impression  Channel3_impression  \
0             670235.0                  0.0   
1             745025.0             226872.0   
2             786262.0             743321.0   
3             190449.0             400702.0   
4             584113.0             346168.0   

   competitor_activity_score_control  sentiment_score_control  Channel0_spend  \
0                          -0.783350                 3.036792       15482.038   
1                          -1.834407                -4.244738       10420.116   
2                          -1.995511                 0.200945       14304.095   
3                          -4.925971                -1.542391       12784.685   
4                          -4.541369                 0.668494       11974.415   

   Channel1_spend  Channel2_spend  Channel3_spend  conversions  \
0       42.681618       8004.1030          0.0000   12530976.0   
1     8257.458000       8897.2630       2668.1526    4926880.5   
2     3770.548600       9389.7250       8741.9060   10300557.0   
3     6055.552700       2274.3865       4712.4990    7398028.5   
4    12086.009000       6975.6143       4071.1460    9409540.0   

   revenue_per_conversion  population  Channel3_reach  Channel3_frequency  
0                0.035213    487878.0             0.0            0.000000  
1                0.035026    487878.0        181472.0            1.250176  
2                0.034529    487878.0        381866.0            1.946549  
3                0.035430    487878.0        273012.0            1.467708  
4                0.034637    487878.0        247506.0            1.398625

<a id="philosophy"></a>
## **Core Philosophy: Geo-Level Modeling & Data Volume**

Meridian can run on national-level or geographical (geo) level data, but **geo-level modeling is strongly recommended**. A national model is simply treated as a special case where there is only one geo.

Structuring your data at a geographical level (e.g., State, Province, or Designated Market Area) provides significant statistical advantages:

* **Increased Statistical Power:** Instead of having one observation per week for the entire country, you have multiple observations (one for each geo). This increases the effective sample size, leading to more precise and reliable measurement of media effects.
* **Better Estimation of Trends & Seasonality:** With more data points per time period, the model can more accurately distinguish between a true marketing impact and underlying temporal patterns like seasonality.
* **Flexibility:** Geo-level data allows the model to "borrow strength" across regions, making it more robust even when some channels have sparse data in certain geos.

**What if I only have national-level data for some channels?**
This is a common scenario. It's often better to impute or allocate the national-level data across your geos than to default to a fully national model. A standard approach is to distribute the national metric (like spend or impressions) based on each geo's share of the national population.


<a id="national-structure"></a>
### **National-Level Data Structure**

With a national model, you have a single observation for the entire country for each time period.

**Example: National Data**
Notice there is only one row per `time` stamp.

| time | Channel0_spend | conversions |
| :--- | :--- | :--- |
| 2024-01-01 | 550,000 | 8,200 |
| 2024-01-08 | 575,000 | 8,550 |
| 2024-01-15 | 560,000 | 8,310 |

Over two years, this structure would give you only 104 data points (52 weeks x 2 years) to analyze.

<a id="geo-structure"></a>
### **Geo-Level Data Structure**

Structuring your data at a geographical level (e.g., State, Province, or [Designated Market Area](https://en.wikipedia.org/wiki/Media_market)) provides multiple observations for each time period.

**Example: Geo-Level Data**
For the same period, you now have multiple rows per `time` stamp—one for each region.

| geo | time | Channel0_spend | conversions | population |
| :--- | :--- | :--- | :--- | :--- |
| CA | 2024-01-01 | 80,000 | 1,150 | 39,200,000 |
| TX | 2024-01-01 | 55,000 | 750 | 29,500,000 |
| NY | 2024-01-01 | 45,000 | 680 | 19,800,000 |
| CA | 2024-01-08 | 82,000 | 1,210 | 39,200,000 |
| TX | 2024-01-08 | 58,000 | 790 | 29,500,000 |
| NY | 2024-01-08 | 48,000 | 710 | 19,800,000 |

If you model 50 geos over two years, you now have 50 * 104 = **5,200 data points**. This massive increase in data provides significant statistical advantages:

* **Increased Statistical Power:** More data leads to more precise and reliable measurement of media effects. The model has more evidence to learn from, resulting in tighter credible intervals around your ROI estimates.
* **Better Estimation of Trends & Seasonality:** With multiple observations per week, the model can more accurately separate the impact of a national holiday or seasonal trend from the impact of your marketing. This is captured by the time-varying intercept, denoted as $\mu_t$.
* **Flexibility:** Geo-level data allows the model to "borrow strength" across regions. It can learn about a channel's general effectiveness from high-spend geos and apply that learning to lower-spend geos, making the model more robust.


<a id="schema"></a>
## **The Meridian Input Schema**

All of your prepared data must be consolidated into a single CSV file with a "wide" format. This means each row represents one time period (week) for one geo, and the columns represent all the different variables.

The table below details the fields required for different model configurations.

| Field Name | Data Type | Description | Geo Model | National Model | Reach & Frequency | Sample Value |
| :--- | :--- | :--- | :---: | :---: | :---: | :--- |
| **`geo`** | String | A categorical identifier for the geographical region. | ✔️ | | - | `'CA'` |
| **`time`** | Date | The start date of the week (YYYY-MM-DD format). | ✔️ | ✔️ | - | `'2024-01-01'` |
| **`Channel{i}_impression`** | Float | Total impressions for media channel `i`. | ✔️ | ✔️ | - | `1500000.0` |
| **`Channel{i}_spend`** | Float | Total spend for media channel `i`. | ✔️ | ✔️ | - | `25000.50` |
| **`Channel{i}_reach`** | Float | Unique reach for a channel `i` (if using RF data). | ✔️ | ✔️ | ✔️ | `800000.0` |
| **`Channel{i}_frequency`**| Float | Average frequency for a channel `i` (if using RF data). | ✔️ | ✔️ | ✔️ | `1.875` |
| **`Organic{j}_impression`**| Float | Impressions/volume for organic channel `j`. | ✔️ | ✔️ | - | `50000.0` |
| **`Promo_control`** | Float | A numeric value representing promotional activity. | ✔️ | ✔️ | - | `1.0` |
| **`Gqv_control`** | Float | Indexed Google Query Volume as a control variable. | ✔️ | ✔️ | - | `85.4` |
| **`conversions`** | Float | The target outcome metric (e.g., sales, sign-ups). | ✔️ | ✔️ | - | `1250.0` |
| **`revenue_per_conversion`**| Float | The average revenue for each conversion. | ✔️ | ✔️ | - | `49.99` |
| **`population`** | Float | The total population of the specified geo. | ✔️ | | - | `39200000.0` |

## A Closer Look at Key Input Variables
<a id="key-variables"></a>

To prepare your data correctly, it's important to understand what each key variable represents. These are the columns that will form your final input file.

### Media Metrics: Spend, Impressions, Organic Channels, Reach & Frequency
<a id="media-metrics"></a>

These variables quantify your paid marketing efforts for each channel.

* **`spend`**: The total cost of a marketing activity for the given week and geo.
    * **Schema Name**: `Channel{i}_spend`
    * **Data Type**: Float (e.g., `15482.03`)

* **`impressions`**: The total number of times your advertisements were displayed on a screen.
    * **Schema Name**: `Channel{i}_impression`
    * **Data Type**: Integer or Float (e.g., `1392518.0`)

* **`reach`**: The number of unique individuals exposed to your ads at least once in a given period. [Within the MMM Data Platform, **this data is specifically provided for YouTube campaigns** and is essential for advanced modeling.](https://developers.google.com/meridian/docs/basics/using-mmm-data-platform)
    * **Schema Name**: `Channel{i}_reach`
    * **Data Type**: Integer or Float (e.g., `850000.0`)

* **`frequency`**: The average number of times an individual was exposed to your ads from a specific channel. This is provided alongside **`reach`** for YouTube campaigns to enable Reach & Frequency (RF) modeling, which offers a more nuanced view than impressions alone.
    * **Schema Name**: `Channel{i}_frequency`
    * **Data Type**: Float (e.g., `2.75`)

> **A Note on Organic Variables:**
> Organic channels (like SEO, email, or non-paid social posts) are also key inputs. They are represented by a volume metric (e.g., `Organic{j}_impression`) but crucially have a **`spend` of $0**. Including them is vital for the model to correctly attribute sales and not inflate the ROI of your paid channels.

### Control Variables: Promotions, competitors, brand sentiment, and more
<a id="control-variables"></a>


Think of control variables as factors you don't directly control but that have a real impact on your sales. Their job is to help the model "control for" outside influences, ensuring that your marketing channels get the proper amount of credit—no more, no less. Without them, your model might give your marketing credit for a sales spike that was actually caused by a holiday, a competitor's misstep, or a viral social media trend.

Based on the sample data, here are some key examples:

* **`Promo`**: A numerical score representing your company's promotional activity. This helps the model distinguish between a sales lift from a 20% off coupon versus a lift from advertising.
    * **Example**: You could use a binary system (`1` for a week with a promotion, `0` for a week without) or a weighted score (e.g., `0.5` for a small discount, `1.0` for a major sale).
    * **Data Type**: Float or Integer

* **`competitor_sales_control`**: A variable that tracks the activity or impact of your competitors. This could be their estimated sales, their media spend, or a score representing the intensity of their promotional activity.
    * **Example**: A value of `1.5` could mean a competitor launched a major new product that week, while `-1.0` could mean they had a supply chain issue.
    * **Data Type**: Float

* **`sentiment_score_control`**: A metric that captures the overall public mood or opinion about your brand, often derived from social media listening tools or surveys. A positive shift in sentiment can lift sales on its own.
    * **Example**: A score ranging from `-5` (very negative) to `+5` (very positive).
    * **Data Type**: Float

* **`GQV`**: Stands for **Google Query Volume**, which is provided via MMM Data Platform. This is an indexed value representing the search interest in keywords relevant to your brand or category. It controls for organic interest shifts that could affect sales independently of your marketing.
    * **Example**: An increase in search volume for "beach towels" in late spring is a seasonal trend that GQV helps account for.
    * **Data Type**: Float
    * **Note**: This is currently not included in the samples.


### Frequently Asked Questions about Control Variables

#### **What happens if I don't include them at all?**

Your model's results will likely be wrong, and your budget decisions will be based on flawed information. This is called **omitted variable bias**.

**Example:** Imagine you run a big advertising campaign during the week of Black Friday. Sales skyrocket. If you don't include a control variable for the Black Friday holiday promotion, the model will attribute **100% of that sales lift to your ad campaign**. Your calculated ROI will be massively inflated, and you might be fooled into thinking the campaign was far more effective than it actually was.

#### **How many control variables are too many?**

There is no magic number, but it's a trade-off.

* **Too Few:** You risk the omitted variable bias described above.
* **Too Many:** You risk **overfitting** the model. This means the model starts to "memorize" the random noise in your historical data instead of learning the true, underlying business drivers. An overfit model looks great on past data but is poor at predicting future results.

**A good rule of thumb is to start with 3-5 of the most important, well-understood external factors.** Only add more if you have a strong, logical reason why they would impact your business and can see that they improve the model's predictive accuracy.

### Outcome Metrics: Conversions & Revenue
<a id="outcome-metrics"></a>

These variables define the business goal you are trying to optimize.

* **`conversions`**: This is your primary target variable. It represents the total number of desired actions completed, such as sales, sign-ups, or leads.
    * **Data Type**: Integer or Float (e.g., `12530976.0`)

* **`revenue_per_conversion`**: The average revenue generated by a single conversion. Including this optional field allows Meridian to directly calculate financial metrics like Return on Investment (ROI).
    * **Data Type**: Float (e.g., `0.0352`)

### Geo-Specific Variable: Population
<a id="geo-variables"></a>

This variable is essential for geo-level modeling.

* **`population`**: The total population of the specific geographical area. This is crucial for weighting each region and is used when allocating national-level spend across different geos.
    * **Data Type**: Integer or Float (e.g., `487878.0`)

<a id="samples"></a>
## **Sample Meridian Input Files**

To make this concrete, here are several examples of what the final, aggregated weekly data should look like. The samples are also available on GitHub.

**Example 1: Weekly Geo-Level Data**




In [ ]:
# Stores the web address (URL) of the raw CSV data file in a variable named 'url'.
geo_media_url = "https://raw.githubusercontent.com/google/meridian/main/meridian/data/simulated_data/csv/geo_all_channels.csv"

# Uses the pandas library (pd) to read the data from the URL and load it into a DataFrame named 'df'.
geo_media_df = pd.read_csv(geo_media_url)

# Displays the first 5 rows of the DataFrame ('df.head(5)') to give a quick preview of the data structure and its contents.
display(geo_media_df.head(5))

Unnamed: 0   geo        time  Channel0_impression  Channel1_impression  \
0           0  Geo0  2021-01-25               280668                    0   
1           1  Geo0  2021-02-01               366206               182108   
2           2  Geo0  2021-02-08               197565               230170   
3           3  Geo0  2021-02-15               140990                66643   
4           4  Geo0  2021-02-22               399116               164991   

   Channel2_impression  Channel3_impression  Channel4_impression  \
0                    0               470611               108010   
1                19825               527702               252506   
2                    0               393618               184061   
3                    0               326034               201729   
4                    0               381982               153973   

   competitor_sales_control  sentiment_score_control  Channel0_spend  \
0                 -1.338765                 0.115581       2058.0608   
1                  0.893645                 0.944224       2685.2874   
2                 -0.284549                -1.290579       1448.6895   
3                 -1.034740                -1.084514       1033.8406   
4                 -0.319276                -0.017503       2926.6072   

   Channel1_spend  Channel2_spend  Channel3_spend  Channel4_spend  \
0         0.00000         0.00000       3667.3965        841.6044   
1      1755.74540       147.31808       4112.2974       1967.5044   
2      2219.12230         0.00000       3067.4023       1434.1870   
3       642.52057         0.00000       2540.7310       1571.8545   
4      1590.71640         0.00000       2976.7249       1199.7440   

   Organic_channel0_impression     Promo  conversions  revenue_per_conversion  \
0                        97320  0.000000    1954576.8                0.020055   
1                       201441  0.000000    2064249.6                0.020103   
2                            0  0.683819    2086382.8                0.019929   
3                            0  1.289055    2826431.5                0.019987   
4                            0  0.227739    3551929.2                0.020000   

   population  
0   136670.94  
1   136670.94  
2   136670.94  
3   136670.94  
4   136670.94

**Example 2: Weekly Geo-Level Data with Reach & Frequency**

In this example dataset, Reach and Frequency are provided for the Channel3.

In [ ]:
# Stores the web address (URL) of the raw CSV data file in a variable named 'url'.
geo_media_rf_url = "https://raw.githubusercontent.com/google/meridian/main/meridian/data/simulated_data/csv/geo_media_rf.csv"

# Uses the pandas library (pd) to read the data from the URL and load it into a DataFrame named 'df'.
geo_media_rf_df = pd.read_csv(geo_media_rf_url)

# Displays the first 5 rows of the DataFrame ('df.head(5)') to give a quick preview of the data structure and its contents.
display(geo_media_rf_df.head(5))

geo        time  Channel0_impression  Channel1_impression  \
0  Geo0  2021-01-25            1392518.0               3733.0   
1  Geo0  2021-02-01             937228.0             722210.0   
2  Geo0  2021-02-08            1286569.0             329778.0   
3  Geo0  2021-02-15            1149907.0             529628.0   
4  Geo0  2021-02-22            1077028.0            1057061.0   

   Channel2_impression  Channel3_impression  \
0             670235.0                  0.0   
1             745025.0             226872.0   
2             786262.0             743321.0   
3             190449.0             400702.0   
4             584113.0             346168.0   

   competitor_activity_score_control  sentiment_score_control  Channel0_spend  \
0                          -0.783350                 3.036792       15482.038   
1                          -1.834407                -4.244738       10420.116   
2                          -1.995511                 0.200945       14304.095   
3                          -4.925971                -1.542391       12784.685   
4                          -4.541369                 0.668494       11974.415   

   Channel1_spend  Channel2_spend  Channel3_spend  conversions  \
0       42.681618       8004.1030          0.0000   12530976.0   
1     8257.458000       8897.2630       2668.1526    4926880.5   
2     3770.548600       9389.7250       8741.9060   10300557.0   
3     6055.552700       2274.3865       4712.4990    7398028.5   
4    12086.009000       6975.6143       4071.1460    9409540.0   

   revenue_per_conversion  population  Channel3_reach  Channel3_frequency  
0                0.035213    487878.0             0.0            0.000000  
1                0.035026    487878.0        181472.0            1.250176  
2                0.034529    487878.0        381866.0            1.946549  
3                0.035430    487878.0        273012.0            1.467708  
4                0.034637    487878.0        247506.0            1.398625

**Example 3: Weekly National-Level Data**

In some cases, depending on your scenario, you may not have the ability to construct a geo level dataset. In cases like this, you can still aggregate your data on a national basis and have Meridian run.

In [ ]:
# Stores the web address (URL) of the raw CSV data file in a variable named 'url'.
national_all_channels_url = "https://raw.githubusercontent.com/google/meridian/main/meridian/data/simulated_data/csv/national_all_channels.csv"

# Uses the pandas library (pd) to read the data from the URL and load it into a DataFrame named 'df'.
national_all_channels_df = pd.read_csv(national_all_channels_url)

# Displays the first 5 rows of the DataFrame ('df.head(5)') to give a quick preview of the data structure and its contents.
display(national_all_channels_df.head(5))

time  conversions  revenue_per_conversion  Channel0_impression  \
0  2021-01-25  386300926.3                0.020000             44340630   
1  2021-02-01  355303931.3                0.019990             47391940   
2  2021-02-08  438308889.8                0.019993             28960185   
3  2021-02-15  418612413.7                0.019986             38126016   
4  2021-02-22  355785298.0                0.019969             46890794   

   Channel1_impression  Channel2_impression  Channel3_impression  \
0             10537596              6501889             66738506   
1             30857600              7966320             79000018   
2             23144749              4637739             68442555   
3             10667077              3029637             70601091   
4             17925961             10486840             81842660   

   Channel4_impression  Channel0_spend  Channel1_spend  Channel2_spend  \
0             53026011    325137.56830   101595.404270     48315.04518   
1             27592214    347511.96809   297505.263890     59197.12148   
2             20057461    212357.01945   223143.885196     34462.68770   
3             37246623    279567.52290   102843.758830     22513.00335   
4             56259530    343837.20920   172828.342330     77926.91530   

   Channel3_spend  Channel4_spend  Organic_channel0_impression  \
0     520082.5242   413173.984050                     24028870   
1     615634.5403   214996.093200                     29188267   
2     533361.9165   156285.964472                     12051369   
3     550183.0380   290222.395240                      9595212   
4     637786.8024   438369.280950                      9288846   

   competitor_sales_control  sentiment_score_control     Promo  
0                 -0.032438                -0.342020  0.272211  
1                  0.312002                 0.807150  0.151264  
2                 -1.243697                -0.152005  0.596395  
3                 -0.019225                -0.377523  0.341224  
4                  0.344244                 0.034181  0.366063

**Example 4: Weekly National-Level Data with Reach & Frequency**

Even in a scenario when you have data aggregated on a national basis, you should still be able to provide reach & frequency data for a certain channel.

In [ ]:
# Stores the web address (URL) of the raw CSV data file in a variable named 'url'.
national_media_rf_url = "https://raw.githubusercontent.com/google/meridian/main/meridian/data/simulated_data/csv/national_media_rf.csv"

# Uses the pandas library (pd) to read the data from the URL and load it into a DataFrame named 'df'.
national_media_rf_df = pd.read_csv(national_media_rf_url)

# Displays the first 5 rows of the DataFrame ('df.head(5)') to give a quick preview of the data structure and its contents.
display(national_media_rf_df.head(5))

time  conversions  revenue_per_conversion  Channel0_impression  \
0  2021-01-25    224655360                0.034894             29203792   
1  2021-02-01    158389250                0.035067             24655420   
2  2021-02-08    216225630                0.034893             31907648   
3  2021-02-15    203125630                0.035203             28151956   
4  2021-02-22    222105420                0.035005             23894308   

   Channel1_impression  Channel2_impression  Channel3_impression  \
0              8887214              6888124               913520   
1             13865920             13265415              6044235   
2             19497852             16535126             11904464   
3              9928580              5394670              4685435   
4             14377868             14984190             14355571   

   Channel0_spend  Channel1_spend  Channel2_spend  Channel3_spend  \
0       324688.25       101612.82       82259.586       10743.550   
1       274119.34       158537.33      158418.690       71083.875   
2       354749.75       222930.58      197466.340      140003.730   
3       312993.90       113519.38       64424.410       55103.562   
4       265657.30       164390.75      178944.700      168830.230   

   Channel3_reach  Channel3_frequency  competitor_activity_score_control  \
0          728787            1.253480                          -0.416371   
1         4242109            1.424818                          -0.846952   
2         6826023            1.743982                          -1.098846   
3         3200847            1.463811                          -0.779951   
4         7487385            1.917301                          -0.982015   

   sentiment_score_control  
0                 0.683978  
1                -1.381805  
2                 0.601687  
3                -0.330715  
4                 0.638006

With this foundation, you are now ready for Part 2, where we'll discuss how to transform raw data from sources like [MMM Data Platform](https://developers.google.com/meridian/docs/basics/using-mmm-data-platform) into these precise formats.